policy/src/main/java/org/apache/brooklyn/policy/ha/ElectPrimaryEffector.java

```diff
@@ -0,0 +1,440 @@
+                    ServiceProblemsLogic.updateProblemsIndicator(entity(), "primary", e);
+                    ServiceStateLogic.setExpectedStateRunningWithErrors(entity());
+                    Exceptions.propagate(e);
+                }
+    
+                return MutableMap.of("code", ResultCode.NEW_PRIMARY_ELECTED, "message", "New primary found", "primary", newPrimary);
+                
+            } catch (Exception e) {
+                Exceptions.propagateIfFatal(e);
+
+                if (Entities.isNoLongerManaged(entity())) {
+                    // ignore errors if shutting down
+                    return "<no-longer-managed>";
+                }
+
+                Lifecycle expected = ServiceStateLogic.getExpectedState(entity());
+                if (expected==Lifecycle.RUNNING || expected==Lifecycle.STARTING) {
+                    // including SelectionModeStrictFailed
+                    log.warn("Error electing new primary at "+entity()+": "+Exceptions.collapseText(e));
+                    ServiceProblemsLogic.updateProblemsIndicator(entity(), "primary", "Error electing primary: "+
+                        Exceptions.collapseText(e));
+                    entity().sensors().set(Attributes.SERVICE_STATE_ACTUAL, Lifecycle.ON_FIRE);
+                    
+                    throw Exceptions.propagateAnnotated("Error electing primary (when "+expected.toString().toLowerCase()+")", e);
+                }
+                
+                throw Exceptions.propagateAnnotated("Error electing primary (when not starting/running)", e);
+            }
+        }
+
+        @SuppressWarnings({ "unchecked", "rawtypes" })
+        protected void promoteAndDemote(ConfigBag params, Entity oldPrimary, Entity newPrimary) {
+            params.configureStringKey("oldPrimary", oldPrimary);
+            params.configureStringKey("newPrimary", newPrimary);
+            MutableList<Task> tasks = MutableList.<Task>of();
+            
+            if (newPrimary!=null) tasks.append(Tasks.create("promote "+newPrimary, new Promote(params)));
+            else tasks.append(Tasks.warning("No new primary; nothing to promote", null, false));
+            
+            if (oldPrimary!=null) tasks.append(Tasks.create("demote "+oldPrimary, new Demote(params)));
+            else tasks.append(Tasks.warning("No old primary; nothing to demote", null, false));
+            
+            log.debug("Running "+tasks);
+            List<?> result = DynamicTasks.queue(Tasks.parallel("promote/demote", (List<Task<?>>)(List) tasks)).getUnchecked();
+            log.debug("Ran "+tasks+", results: "+result);
+        }
+        
+        private Entity getCurrentActive(ConfigBag params) {
+            return entity().getAttribute(Sensors.newSensor(Entity.class, params.get(PRIMARY_SENSOR_NAME)));
+        }
+
+        protected class CheckPrimaries implements Callable<Entity> {
```

Comment:
Would be nice to be able to unit test this logic directly (rather than just in bigger tests that also include the promotion). It is very complicated with many different conditions and varying behaviour (e.g. the `delayForBest` is tested in `ElectPrimaryTest`).

It's hard to understand, particularly the subtleties of the different behaviours. I had to run it in the debugger to convince myself I understood it well enough.

A couple of thoughts on that: if we split it into separate election and promotion effectors, maybe it would be easier to unit test (e.g. fewer permutations, because they can be tested separately). Could/should we refactor slightly to call this method directly from tests?

Reply:
made this `@VisibleForTesting public static` and have one call to it from tests now